packages/vite/src/node/plugins/importAnalysis.ts

```diff
@@ -409,268 +409,288 @@ export function importAnalysisPlugin(config: ResolvedConfig): Plugin {
+              !isDynamicImport &&
+              isLocalImport &&
+              config.server.preTransformRequests
             ) {
-              needQueryInjectHelper = true
-              str().overwrite(
-                start,
-                end,
-                `__vite__injectQuery(${url}, 'import')`,
-                { contentOnly: true },
+              // pre-transform known direct imports
+              // These requests will also be registered in transformRequest to be awaited
+              // by the deps optimizer
+              const url = removeImportQuery(hmrUrl)
+              server.transformRequest(url, { ssr }).catch((e) => {
+                if (e?.code === ERR_OUTDATED_OPTIMIZED_DEP) {
+                  // This are expected errors
+                  return
+                }
+                // Unexpected error, log the issue but avoid an unhandled exception
+                config.logger.error(e.message)
+              })
+            }
+          } else if (!importer.startsWith(clientDir)) {
+            if (!isInNodeModules(importer)) {
+              // check @vite-ignore which suppresses dynamic import warning
+              const hasViteIgnore = hasViteIgnoreRE.test(
+                // complete expression inside parens
+                source.slice(dynamicIndex + 1, end),
               )
+              if (!hasViteIgnore) {
+                this.warn(
+                  `\n` +
+                    colors.cyan(importerModule.file) +
+                    `\n` +
+                    colors.reset(generateCodeFrame(source, start)) +
+                    colors.yellow(
+                      `\nThe above dynamic import cannot be analyzed by Vite.\n` +
+                        `See ${colors.blue(
+                          `https://github.com/rollup/plugins/tree/master/packages/dynamic-import-vars#limitations`,
+                        )} ` +
+                        `for supported dynamic import formats. ` +
+                        `If this is intended to be left as-is, you can use the ` +
+                        `/* @vite-ignore */ comment inside the import() call to suppress this warning.\n`,
+                    ),
+                )
+              }
+            }
+
+            if (!ssr) {
+              const url = rawUrl.replace(cleanUpRawUrlRE, '').trim()
+              if (
+                !urlIsStringRE.test(url) ||
+                isExplicitImportRequired(url.slice(1, -1))
+              ) {
+                needQueryInjectHelper = true
+                str().overwrite(
+                  start,
+                  end,
+                  `__vite__injectQuery(${url}, 'import')`,
+                  { contentOnly: true },
+                )
+              }
             }
           }
-        }
-      }
+        }),
+      )
+
+      const acceptedUrls = mergeAcceptedUrls(orderedAcceptedUrls)
+      const acceptedExports = mergeAcceptedUrls(orderedAcceptedExports)
```

Comment:
Merged after the Promise.all so these arrays are always ordered in the same way for each file. This isn't strictly needed but should help when debugging.